### ISSUE TITLE
Game video output not recognizing scaled displays. 

### ISSUE THREAD

**Issue by evertvorster** [vegastrike/Vega-Strike-Engine-Source#884]
One of my monitors is fairly high resolution, and so it is scaled so that it is usable. 
Unfortunately this causes issues with VegaStrike which is strictly pixel based. 

SDL has a fairly straightforward way of dealing with this that should not impact non-scaled displays:
https://wiki.libsdl.org/SDL2/SDL_HINT_WINDOWS_DPI_SCALING

**Comment by royfalk** [comment]
What happens if you try 1024x768 (or some other low resolution) in full screen?

**Comment by royfalk** [comment]
OK. See if you can find a reference in SDL as well as a discussion or code snippet to implement this. If you can find it, I'll implement this. Contingent on approval by @BenjamenMeyer.

**Comment by royfalk** [comment]
OK. That looks good. 
Let's give Ben a chance to think about this.

**Comment by BenjamenMeyer** [comment]
1. We probably need to take a stance about AI generators (Copilot, Gemini, ChatGPT, etc); most open source projects have bans on them (namely due to liability issues). So while the comment above is helpful and informative we should use it as a reference and not directly to code. We probably should open a separate discussion about AI generators.
2. We should have something in VS Settings sooner than later. I'm good with that. Ultimately I think we should integrate VS Settings into the game itself so you don't have to leave the game to change things.
3. We should probably scale the fonts separate from the resolution so we can support small/medium/large/extra large fonts to avoid issues with either fonts being too small or too large. Mapping it to the resolution does not necessarily resolve that.

**Comment by royfalk** [comment]
> 1. We probably need to take a stance about AI generators (Copilot, Gemini, ChatGPT, etc); most open source projects have bans on them (namely due to liability issues). So while the comment above is helpful and informative we should use it as a reference and not directly to code. We probably should open a separate discussion about AI generators.

That's a good point. I wasn't about to plug this in as is anyhow. However, I did not consider this as I've been out of a job for a while now. No doubt my old job would have had a training session regarding this.

>     2. We should have something in VS Settings sooner than later. I'm good with that. Ultimately I think we should integrate VS Settings into the game itself so you don't have to leave the game to change things.

The big block here is the GUI. I don't want to tackle any GUI development using the home-grown stuff they wrote. Once we figure out something else, this should be reasonably easy.
 
>     3. We should probably scale the fonts separate from the resolution so we can support small/medium/large/extra large fonts to avoid issues with either fonts being too small or too large. Mapping it to the resolution does not necessarily resolve that.

I'm running the game in windowed 2k and the fonts are tiny. I just live with it until we can get to the GUI side of things.

**Comment by BenjamenMeyer** [comment]
> > 1. We probably need to take a stance about AI generators (Copilot, Gemini, ChatGPT, etc); most open source projects have bans on them (namely due to liability issues). So while the comment above is helpful and informative we should use it as a reference and not directly to code. We probably should open a separate discussion about AI generators.
> 
> That's a good point. I wasn't about to plug this in as is anyhow. However, I did not consider this as I've been out of a job for a while now. No doubt my old job would have had a training session regarding this.
> 
> >

Yeah, my current employer introduced me to sum of the intricacies of this. I have a Copilot licences for work; but the liability for the license only extends to who paid for it. There's also not much protection on it from cross-using licenses so...I'll start a separate discussion on the topic.

**Comment by stephengtuggy** [comment]
@evertvorster I started working on this today. I found the same SDL_HINT_... setting, independently, and started experimenting with it on my Windows box. It seems to work pretty well.

**Comment by evertvorster** [comment]
@stephengtuggy 
This is quite the nuanced issue, especially since we are running VS on multiple platforms.

Ideally, top level, when the game is launched, it should read the available screen resolutions for each monitor. 

There should then be some sanity check to see whether the screen and resolution in the config is actually available. 
If it is, then happy days! Launch the game with the requested config. 

If it is not, ideally the game should provide some way of setting new values for the config, or at the very least not launch at all with an error message indicating that the requested resolutions are not available.

**Comment by evertvorster** [comment]
Where the nuances come in is that there are at least two types of scaling. 
What is done by the system should be irrelevant to the game. 
The system will report resolutions available to VS, and that should be what the game uses. No second guessing of what the system provides. 

What is left, of course, is the scaling done by VS. 
The user has the option of windowed or not. Windowed is kind of easy, just make a window of the requested resolution. At the moment, VS does not support scaling dynamically, so the created window should not be resize- able.  If SDL supports dynamically adjusting scaling to a window, this might allow us to resize the window after launching the game. 

For full screen, you can either have just the native resolution of the screen, or some smaller resolution that should be scaled up to the native resolution by SDL. 
Ideally, preserving aspect ratio should be a config option. I can't imagine wanting to play on a horizontally stretched display, but there may be some weird non-square pixel displays out there.. who knows?

One last thing to consider is the text height. It is a configurable option in the old config file, and I have used it to get readable text on high resolution monitors. Unfortunately in the game menus a bigger text height does not translate into a higher field for the text to go in to. and so text on different lines tend to overlap.  For the rest of the game it is just fine.

**Comment by stephengtuggy** [comment]
@evertvorster yeah, that sounds about right for what is needed in the longer-term. In the short term, however, I am starting with something simpler: just telling the operating system that the application is high-DPI-aware, using that SDL hint you mentioned, `SDL_HINT_WINDOWS_DPI_SCALING`.

**Comment by stephengtuggy** [comment]
Up until now, I have been manually updating the compatibility properties for the EXE in File Explorer, overriding the high-DPI setting to tell Windows that the application will handle it. With this setting, that is no longer necessary, which I think is nice.

**Comment by AnotherCommander** [comment]
The SDL hint is probably fine, but the way recommended by Microsoft for handling scaling is to include the dpiAware and dpiAwareness keys in the application manifest embedded in the executable like at the bottom of this example:

```
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0">
	<compatibility xmlns="urn:schemas-microsoft-com:compatibility.v1"> 
		<application> 
			<!-- Windows 10 --> 
			<supportedOS Id="{8e0f7a12-bfb3-4fe8-b9a5-48fd50a15a9a}"/>
			<!-- Windows 8.1 -->
			<supportedOS Id="{1f676c76-80e1-4239-95bb-83d0f6d0da78}"/>
			<!-- Windows Vista -->
			<supportedOS Id="{e2011457-1546-43c5-a5fe-008deee3d3f0}"/> 
			<!-- Windows 7 -->
			<supportedOS Id="{35138b9a-5d96-4fbd-8e2d-a2440225f93a}"/>
			<!-- Windows 8 -->
			<supportedOS Id="{4a2f28e3-53b9-4441-ba9c-d69d4a4a6e38}"/>
		</application> 
	</compatibility>
	<application xmlns="urn:schemas-microsoft-com:asm.v3">
		<windowsSettings>
			<dpiAware xmlns="http://schemas.microsoft.com/SMI/2005/WindowsSettings">true/PM</dpiAware>
			<dpiAwareness xmlns="http://schemas.microsoft.com/SMI/2016/WindowsSettings">PerMonitorV2, PerMonitor</dpiAwareness>
		</windowsSettings>
	</application>
</assembly>
```
If you embed this in the exe there will be no more DPI scaling issues and it works on all versions of Windows.

**Comment by stephengtuggy** [comment]
@AnotherCommander that's good to know, thanks.

**Comment by evertvorster** [comment]
This issue has gone away.
